### PR TITLE
feat(core/dfn): support "no-export" class

### PR DESCRIPTION
**Pull request by marcoscaceres**
Support for `<dfn class="no-export">`, giving control to Editors over things they might want to keep private.

### CODE REVIEW

File: src/core/dfn.js [hunk 1]
```diff
@@ -86,32 +86,64 @@ function computeTypeAndExport(dfn, linkingText) {
...
+    // Error if we have both exports and no exports.
+    case dfn.matches(".export.no-export"): {
+      const msg = docLink`Declares both "${"[no-export]"}" and "${"[export]"}" css class.`;
+      const hint = "Please use only one.";
+      showError(msg, name, { elements: [dfn], hint });
+      break;
+    }
+
+    // No export wins
+    case isNoExport(dfn):
```
Comment (sidvishnoi) [inline src/core/dfn.js:123]: Let's inline this for consistency

File: src/core/dfn.js [hunk 2]
```diff
@@ -86,32 +86,64 @@ function computeTypeAndExport(dfn, linkingText) {
...
+        const msg = docLink`Declares ${"[no-export]"} CSS class, but also has a "${"[data-export]"}" attribute.`;
+        const hint = "Please chose only one.";
+        showError(msg, name, { elements: [dfn], hint });
+        delete dfn.dataset.export;
+      }
+      dfn.dataset.noexport = "";
+      break;
+
+    // If the author explicitly asked for it to be exported, so let's export it.
+    case dfn.matches(".export") && !dfn.matches("[data-no-export]"):
```
Comment (sidvishnoi) [inline src/core/dfn.js:134]: I'm confused with the usage of `data-no-export` and `data-noexport`

File: tests/spec/core/dfn-spec.js [hunk 3]
```diff
@@ -568,5 +568,64 @@ describe("Core — Definitions", () => {
...
+      expect(idlDfn.dataset.noexport).toBe("");
+      expect(idlDfn.dataset.dfnType).toBe("interface");
+      expect(idlDfn.dataset.export).toBeUndefined();
+
+      // Check validation error
+      const errors = findDfnErrors(doc);
+      expect(errors).toHaveSize(0);
+    });
+
+    it("errors if the dfn declares export and no-export classes", async () => {
```
Comment (sidvishnoi) [inline tests/spec/core/dfn-spec.js:612]:
```suggestion
    it("errors if the dfn declares both export and no-export classes", async () => {
```

File: src/core/dfn.js [hunk 4]
```diff
@@ -86,32 +86,64 @@ function computeTypeAndExport(dfn, linkingText) {
...
+}
 
-  if (!dfn.dataset.dfnType && type) dfn.dataset.dfnType = type;
-  if (shouldExport && !dfn.hasAttribute("data-noexport")) {
-    dfn.dataset.export = "";
+// Deal with export/no export
+function computeExport(dfn) {
+  switch (true) {
+    // Error if we have both exports and no exports.
+    case dfn.matches(".export.no-export"): {
```
[the remark this answers is not in the document]
Comment (marcoscaceres) [reply src/core/dfn.js:115]: humans :) 
```
<dfn class="style-foo bar foo-bar export something no-export"> 
``

File: src/core/dfn.js (hunk 2 above)
Comment (marcoscaceres) [reply src/core/dfn.js:134]: oops... this is why `data-noexport` should have never been.

File: src/core/dfn.js [hunk 5]
```diff
@@ -86,32 +86,64 @@ function computeTypeAndExport(dfn, linkingText) {
...
+        const msg = docLink`Declares ${"[no-export]"} CSS class, but also has a "${"[data-export]"}" attribute.`;
+        const hint = "Please chose only one.";
+        showError(msg, name, { elements: [dfn], hint });
+        delete dfn.dataset.export;
+      }
+      dfn.dataset.noexport = "";
+      break;
+
+    // If the author explicitly asked for it to be exported, so let's export it.
+    case dfn.matches(".export") && !dfn.matches("[data-no-export]"):
```
Comment (marcoscaceres) [inline src/core/dfn.js:134]:
```suggestion
    case dfn.matches(".export") && !dfn.matches("[data-noexport]"):
```

File: src/core/dfn.js [hunk 6]
```diff
@@ -86,32 +86,64 @@ function computeTypeAndExport(dfn, linkingText) {
...
+    // Error if we have both exports and no exports.
+    case dfn.matches(".export.no-export"): {
+      const msg = docLink`Declares both "${"[no-export]"}" and "${"[export]"}" CSS class.`;
+      const hint = "Please use only one.";
+      showError(msg, name, { elements: [dfn], hint });
+      break;
+    }
+
+    // No export wins
+    case isNoExport(dfn):
```
Comment (marcoscaceres) [inline src/core/dfn.js:123]:
```suggestion
    case dfn.matches(".no-export") || dfn.matches("[data-noexport]"):
```

File: src/core/dfn.js [hunk 7]
```diff
@@ -86,29 +86,54 @@ function computeTypeAndExport(dfn, linkingText) {
...
+}
 
-  if (!dfn.dataset.dfnType && type) dfn.dataset.dfnType = type;
-  if (shouldExport && !dfn.hasAttribute("data-noexport")) {
-    dfn.dataset.export = "";
+// Deal with export/no export
+function computeExport(dfn) {
+  switch (true) {
+    // Error if we have both exports and no exports.
+    case dfn.matches(".export.no-export"): {
```
Comment (sidvishnoi) [inline src/core/dfn.js:115]: Maybe can use regular if/else with early return here 😄

Comment (marcoscaceres) [reply src/core/dfn.js:115]: Even with this one, I prefer thinking about them as cases. If it’s all the same, I’d prefer to leave it.

File: src/core/dfn.js [hunk 8]
```diff
@@ -156,7 +181,9 @@ function processAsInternalSlot(title, dfn) {
   }
 
   // Don't export internal slots by default, as they are not supposed to be public.
-  if (!dfn.dataset.hasOwnProperty("export")) dfn.dataset.noexport = "";
+  if (!dfn.matches(".export") && !dfn.matches("[data-export]")) {
```
Comment (sidvishnoi) [inline src/core/dfn.js:184]:
```suggestion
  if (!dfn.matches(".export, [data-export]")) {
```

File: src/core/dfn.js [hunk 9]
```diff
@@ -86,29 +86,54 @@ function computeTypeAndExport(dfn, linkingText) {
...
+    // Error if we have both exports and no exports.
+    case dfn.matches(".export.no-export"): {
+      const msg = docLink`Declares both "${"[no-export]"}" and "${"[export]"}" CSS class.`;
+      const hint = "Please use only one.";
+      showError(msg, name, { elements: [dfn], hint });
+      break;
+    }
+
+    // No export wins
+    case dfn.matches(".no-export") || dfn.matches("[data-noexport]"):
```
Comment (sidvishnoi) [inline src/core/dfn.js:123]:
```suggestion
    case dfn.matches(".no-export, [data-noexport]"):
```

**Review by sidvishnoi — APPROVED** [review]

**Pull request merged**